crates/fluvio-stream-dispatcher/src/metadata/k8.rs

```diff
@@ -0,0 +1,321 @@
+        let k8_spec: S::K8Spec = spec.into_k8();
+
+        let mut input_metadata = if let Some(parent_metadata) = ctx.owner() {
+            debug!("owner exists");
+            let item_name = key.to_string();
+
+            let mut input_metadata = parent_metadata
+                .make_child_input_metadata::<<<S as Spec>::Owner as K8ExtendedSpec>::K8Spec>(
+                    item_name,
+                );
+            // set labels
+
+            if let Some(finalizer) = S::FINALIZER {
+                input_metadata.finalizers = vec![finalizer.to_owned()];
+                for o_ref in &mut input_metadata.owner_references {
+                    o_ref.block_owner_deletion = true;
+                }
+            }
+            input_metadata
+        } else {
+            ctx.item().inner().clone().into()
+        };
+
+        input_metadata.labels = ctx.item().get_labels();
+        input_metadata.annotations = ctx.item().annotations.clone();
+
+        trace!("converted metadata: {:#?}", input_metadata);
+        let new_k8 = InputK8Obj::new(k8_spec, input_metadata);
+
+        debug!("input {:#?}", new_k8);
+
+        K8MetadataClient::apply(self, new_k8).await.map(|_| ())
+    }
+
+    async fn update_spec<S>(&self, metadata: K8MetaItem, spec: S) -> Result<()>
+    where
+        S: K8ExtendedSpec,
+    {
+        debug!("K8 Update Spec: {} key: {}", S::LABEL, metadata.name);
+        trace!("K8 Update Spec: {:#?}", spec);
+        let k8_spec: <S as K8ExtendedSpec>::K8Spec = spec.into_k8();
+
+        trace!("updating spec: {:#?}", k8_spec);
+
+        let k8_input: InputK8Obj<S::K8Spec> = InputK8Obj {
+            api_version: S::K8Spec::api_version(),
+            kind: S::K8Spec::kind(),
+            metadata: metadata.inner().clone().into(),
+            spec: k8_spec,
+            ..Default::default()
+        };
+
+        K8MetadataClient::apply(self, k8_input).await.map(|_| ())
+    }
+
+    async fn update_spec_by_key<S>(
+        &self,
+        key: S::IndexKey,
+        namespace: &NameSpace,
+        spec: S,
+    ) -> Result<()>
+    where
+        S: K8ExtendedSpec,
+    {
+        let meta = K8MetaItem::new(key.to_string(), namespace.to_string());
+        self.update_spec(meta, spec).await
+    }
+
+    async fn update_status<S>(
+        &self,
+        metadata: K8MetaItem,
+        status: S::Status,
+        namespace: &NameSpace,
+    ) -> Result<MetadataStoreObject<S, K8MetaItem>>
+    where
+        S: K8ExtendedSpec,
+    {
+        debug!(
+            key = %metadata.name,
+            version = %metadata.resource_version,
+            "update status begin",
+        );
+        debug!(
+            "K8 Update Status: {} key: {} value: {:?}",
+            S::LABEL,
+            metadata.name,
+            status
+        );
+        trace!("status update: {:#?}", status);
+        let k8_status: <S::K8Spec as K8Spec>::Status = S::convert_status_from_k8(&status);
```

Comment:
is status conversion need to be ref instead of own?

Reply:
Actually, no. Changed it to owned type.